src/libraries/System.Text.Json/src/System/Text/Json/Node/JsonObject.List.cs

```diff
@@ -0,0 +1,386 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text.Json.Serialization.Converters;
+
+namespace System.Text.Json.Node
+{
+    public partial class JsonObject
+    {
+        private const int ListToDictionaryThreshold = 9;
```

Comment:
FWIW I ran benchmarks comparing linear with hashtable lookups in this context:

https://gist.github.com/eiriktsarpalis/1a7916fad982b83fa8b5d1b618ded073

I'm assuming this number was chosen in order to avoid allocating the extra dictionary for most cases but maybe the performance tradeoff for lookups is too substantial to justify that? Perhaps it makes sense to lower the threshold to 4 or 5.

Reply:
Thanks. Note that the number of keys that were present in yours were *2 via `Enumerable.Range(0, 2 * Size).Select`. Not sure if that was intentional.

I also ran local benchmarks and modified yours [here](https://gist.github.com/steveharter/1493ca55fb9135c8753f5346eadd063e) to include the cost of creating the list (which is required even in the dictionary case) and the dictionary. The CPU results are close to around 18 elements, then dictionary is faster but still takes more allocs of course. At 200 elements, the dictionary is 7X faster.

I think the general usage pattern is to read all properties once, so having the benchmark include the overhead of creation and allocs is a fair test.

Reply:
> Not sure if that was intentional

It was, I wanted to average the results with lookups for missing keys. Only the first half of those keys are added to the collections.

Reply:
Other than that, I think it's reasonable to include the allocations in the benchmarks. `9` seems like a reasonable threshold given the numbers.